src/const.rs

```diff
@@ -0,0 +1,131 @@
+                Some(ConstantBool(false)) => constant(cx, otherwise),
+                _ => None,
+            },
+        &ExprLit(ref lit) => Some(lit_to_constant(lit)),
+        &ExprVec(ref vec) => constant_vec(cx, vec),
+        &ExprTup(ref tup) => constant_tup(cx, tup),
+        &ExprUnary(op, ref operand) => constant(cx, operand).and_then(
+            |o| match op {
+                UnNot =>
+                    if let ConstantBool(b) = o {
+                        Some(ConstantBool(!b))
+                    } else { None },
+                UnNeg =>
+                    match o {
+                        &ConstantInt(value, ty) =>
+                            Some(ConstantInt(value, match ty {
+                                SignedIntLit(ity, sign) => 
+                                    SignedIntLit(ity, neg_sign(sign)),
+                                UnsuffixedIntLit(sign) =>
+                                    UnsuffixedIntLit(neg_sign(sign)),
+                                _ => { return None; },
+                            })),
+                        &LitFloat(ref is, ref ty) =>
+                            Some(ConstantFloat(neg_float_str(is), ty)),
+                        &LitFloatUnsuffixed(ref is) => 
+                            Some(ConstantFloatUnsuffixed(neg_float_str(is))),
+                        _ => None,
+                    },
+                UnUniq | UnDeref => o,
+            }),
+        //TODO: add other expressions
+        _ => None,
+    }
+}
+
+fn lit_to_constant(lit: &Lit_) -> Constant {
+    match lit {
+        &LitStr(ref is, style) => ConstantStr(&*is, style),
+        &LitBinary(ref blob) => ConstantBinary(blob.clone()),
+        &LitByte(b) => ConstantByte(b),
+        &LitChar(c) => ConstantChar(c),
+        &LitInt(value, ty) => ConstantInt(value, ty),
+        &LitFloat(ref is, ty) => ConstantFloat(Cow::Borrowed(&*is), ty),
+        &LitFloatUnsuffixed(InternedString) => 
+            ConstantFloatUnsuffixed(Cow::Borrowed(&*is)),
+        &LitBool(b) => ConstantBool(b),
+    }
+}
+
+/// create `Some(ConstantVec(..))` of all constants, unless there is any
+/// non-constant part
+fn constant_vec(cx: &Context, vec: &[&Expr]) -> Option<Constant> {
+    Vec<Constant> parts = Vec::new();
+    for opt_part in vec {
+        match constant(cx, opt_part) {
+            Some(ref p) => parts.push(p),
+            None => { return None; },
+        }
+    }
+    Some(ConstantVec(parts))
+}
+
+fn constant_tup(cx, &Context, tup: &[&Expr]) -> Option<Constant> {
+    Vec<Constant> parts = Vec::new();
+    for opt_part in vec {
+        match constant(cx, opt_part) {
+            Some(ref p) => parts.push(p),
+            None => { return None; },
+        }
+    }
+    Some(ConstantTuple(parts))
+}
+
+/// lookup a possibly constant expression from a ExprPath
+fn fetch_path(cx: &Context, e: &Expr) -> Option<Constant> {
+    if let Some(&PathResolution { base_def: DefConst(id), ..}) =
```

Comment:
I'm not sure where this would be useful. Can you direct me to an example where we could need this? Otherwise, I'm going to implement more Exprs first.